### ISSUE TITLE
Compiler crash on compiling

### ISSUE THREAD

**Issue by rickbutton** [llvm-dcpu16/llvm-dcpu16#161]
Trying to compile this itoa function

```
char* itoa(int value, char* result, int base) {
    // check that the base if valid
    if (base < 2 || base > 36) { *result = '\0'; return result; }

    char* ptr = result, *ptr1 = result, tmp_char;
    int tmp_value;

    do {
        tmp_value = value;
        value /= base;
        *ptr++ = "zyxwvutsrqponmlkjihgfedcba9876543210123456789abcdefghijklmnopqrstuvwxyz" [35 + (tmp_value - value     * base)];
    } while ( value );

    // Apply negative sign
    if (tmp_value < 0) *ptr++ = '-';
    *ptr-- = '\0';
    while(ptr1 < ptr) {
        tmp_char = *ptr;
        *ptr--= *ptr1;
        *ptr1++ = tmp_char;
    }
    return result;
}
```

LLVM throws this error and output when it gets to the function:

```
Assertion failed: (getOperand(0)->getType() == cast<PointerType>(getOperand(1)->getType())->getElementType() && 
"Ptr     must be a pointer to Val type!"), function AssertOK, file /Users/rickbutton/code/dcpu/llvm-dcpu16/lib/VMCore/Instructions.cpp, line 1056.
....
1.  <eof> parser at end of file
2.  src/include/string/itoa.c:6:11: LLVM IR generation of declaration 'itoa'
3.  src/include/string/itoa.c:6:11: Generating code for declaration 'itoa'
4.  src/include/string/itoa.c:6:51: LLVM IR generation of compound statement ('{}')
5.  src/include/string/itoa.c:13:5: LLVM IR generation of compound statement ('{}')
clang-3: error: unable to execute command: Illegal instruction: 4
```

I don't think there is anything syntactically wrong with the function, because I pulled it from [here](http://www.jb.man.ac.uk/~slowe/cpp/itoa.html)

**Comment by ghost** [comment]
Does it compile when using the normal x86 / x86-64 backend/target?

**Comment by Blei** [comment]
I think I found the problem:

``` c
"literal string"[foo]
```

produces a `i8` as return type, but it should produce a `i16` like all other chars... I'm looking into it this evening probably.

**Comment by Blei** [comment]
Of course you guys are welcome to fix it if you have time and want to do it :)

Relevant file (probably): `tools/clang/lib/Sema/SemaExpr.cpp`, ca. line 1200.

**Comment by Blei** [comment]
Fixed by llvm-dcpu16/clang#12
